File: packages/ts/test/messages.test.ts
import { describe, it, expect } from "vitest";
import {
  createRollupMsg,
  submitBatchMsg,
  executeWithdrawalMsg,
  createRdkRegistry,
} from "../src/index";

describe("message builders", () => {
  it("createRollupMsg coerces stake to bigint and sets the type URL", () => {
    const msg = createRollupMsg({
      creator: "qor1c",
      rollupId: "r",
      profile: "defi",
      vmType: "evm",
      stakeAmount: "10000000000",
    });
    expect(msg.typeUrl).toBe("/qorechain.rdk.v1.MsgCreateRollup");
    expect(msg.value.stakeAmount).toBe(10000000000n);
  });

  it("submitBatchMsg parses hex byte fields and defaults optional bytes to empty", () => {
    const msg = submitBatchMsg({
      sequencer: "qor1s",
      rollupId: "r",
      batchIndex: 1,
      stateRoot: "0x0a0b",
      txCount: 5,
      dataHash: new Uint8Array([255]),
    });
    expect(Array.from(msg.value.stateRoot)).toEqual([10, 11]);
    expect(Array.from(msg.value.dataHash)).toEqual([255]);
    expect(msg.value.prevStateRoot.length).toBe(0);
    expect(msg.value.proof.length).toBe(0);
    expect(msg.value.batchIndex).toBe(1n);
  });

  it("executeWithdrawalMsg maps repeated proof bytes", () => {
    const msg = executeWithdrawalMsg({
      submitter: "qor1x",
      rollupId: "r",
      batchIndex: 2,
      withdrawalIndex: 0,
      recipient: "qor1y",
      denom: "uqor",
      amount: 100,
      proof: ["0x01", new Uint8Array([2, 2])],
    });
    expect(msg.value.proof.map((p: Uint8Array) => Array.from(p))).toEqual([[1], [2, 2]]);
  });

  it("the rdk registry can look up every message type", () => {
    const registry = createRdkRegistry();
    for (const typeUrl of [
      "MsgCreateRollup",
      "MsgSubmitBatch",
      "MsgChallengeBatch",
      "MsgResolveChallenge",
      "MsgPauseRollup",
      "MsgResumeRollup",
      "MsgStopRollup",
      "MsgExecuteWithdrawal",
    ]) {
      expect(registry.lookupType(`/qorechain.rdk.v1.${typeUrl}`)).toBeTruthy();
    }
  });
});

File: packages/go/messages_test.go
package rdk

import "testing"

func TestMsgProtoGolden(t *testing.T) {
	g := loadGolden(t)

	createRollup := MsgCreateRollup{
		Creator: "qor1creator", RollupID: "my-rollup", Profile: "defi",
		VmType: "evm", StakeAmount: 10000000000,
	}
	pauseRollup := MsgPauseRollup{Creator: "qor1creator", RollupID: "r1", Reason: "x"}
	submitBatch := MsgSubmitBatch{
		Sequencer: "qor1seq", RollupID: "r", BatchIndex: 7,
		StateRoot: []byte{0x01, 0x02, 0x03}, PrevStateRoot: []byte{0x04, 0x05},
		TxCount: 42, DataHash: []byte{0x09, 0x09}, Proof: []byte{0x08},
		WithdrawalsRoot: []byte{0x07, 0x07, 0x07},
	}
	executeWithdrawal := MsgExecuteWithdrawal{
		Submitter: "qor1sub", RollupID: "r", BatchIndex: 3, WithdrawalIndex: 1,
		Recipient: "qor1dest", Denom: "uqor", Amount: 500,
		Proof: [][]byte{{0x01}, {0x02, 0x02}},
	}

	cases := map[string][]byte{
		"MsgCreateRollup":      createRollup.Marshal(),
		"MsgPauseRollup":       pauseRollup.Marshal(),
		"MsgSubmitBatch":       submitBatch.Marshal(),
		"MsgExecuteWithdrawal": executeWithdrawal.Marshal(),
	}
	for name, got := range cases {
		want, ok := g.MsgProtoHex[name]
		if !ok {
			t.Fatalf("golden missing %s", name)
		}
		if BytesToHex(got) != want {
			t.Errorf("%s proto:\n got %s\nwant %s", name, BytesToHex(got), want)
		}
	}
}

func TestMsgTypeURLs(t *testing.T) {
	cases := map[string]string{
		MsgCreateRollup{}.TypeURL():      "/qorechain.rdk.v1.MsgCreateRollup",
		MsgSubmitBatch{}.TypeURL():       "/qorechain.rdk.v1.MsgSubmitBatch",
		MsgChallengeBatch{}.TypeURL():    "/qorechain.rdk.v1.MsgChallengeBatch",
		MsgResolveChallenge{}.TypeURL():  "/qorechain.rdk.v1.MsgResolveChallenge",
		MsgPauseRollup{}.TypeURL():       "/qorechain.rdk.v1.MsgPauseRollup",
		MsgResumeRollup{}.TypeURL():      "/qorechain.rdk.v1.MsgResumeRollup",
		MsgStopRollup{}.TypeURL():        "/qorechain.rdk.v1.MsgStopRollup",
		MsgExecuteWithdrawal{}.TypeURL(): "/qorechain.rdk.v1.MsgExecuteWithdrawal",
	}
	for got, want := range cases {
		if got != want {
			t.Errorf("type url got %s want %s", got, want)
		}
	}
}
